pkg/networking/subnet_resolver.go

```diff
@@ -263,6 +268,76 @@ func (r *defaultSubnetsResolver) ResolveViaNameOrIDSlice(ctx context.Context, su
 	return resolvedSubnets, nil
 }
 
+func (r *defaultSubnetsResolver) ResolveViaSelector(ctx context.Context, selectors []elbv2api.SubnetSelector, opts ...SubnetsResolveOption) ([]*ec2sdk.Subnet, error) {
+	resolveOpts := defaultSubnetsResolveOptions()
+	resolveOpts.ApplyOptions(opts)
+
+	var subnetIDs []string
+	subnetTags := make(map[string][]string)
+	for _, selector := range selectors {
+		if selector.ID != "" {
+			subnetIDs = append(subnetIDs, selector.ID)
+		}
+		if selector.TagKey != "" {
+			subnetTags[selector.TagKey] = append(subnetTags[selector.TagKey], selector.TagValue)
+		}
+	}
+
+	var resolvedSubnets []*ec2sdk.Subnet
+	if len(subnetIDs) > 0 {
+		req := &ec2sdk.DescribeSubnetsInput{
+			SubnetIds: awssdk.StringSlice(subnetIDs),
+		}
+		subnets, err := r.ec2Client.DescribeSubnetsAsList(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+		if len(subnets) != len(subnetIDs) {
+			return nil, errors.Errorf("couldn't find all subnets, IDs: %v, found: %v", subnetIDs, len(subnets))
+		}
+		resolvedSubnets = append(resolvedSubnets, subnets...)
+	}
+
+	for key, values := range subnetTags {
+		req := &ec2sdk.DescribeSubnetsInput{
+			Filters: []*ec2sdk.Filter{
+				{
+					Name:   awssdk.String("tag:" + key),
+					Values: awssdk.StringSlice(values),
+				},
+				{
+					Name:   awssdk.String("vpc-id"),
+					Values: awssdk.StringSlice([]string{r.vpcID}),
+				},
+			},
+		}
+		subnets, err := r.ec2Client.DescribeSubnetsAsList(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+		if len(subnets) < len(values) {
+			return nil, errors.Errorf("couldn't find all subnets, tag: %v, found: %v", key, len(subnets))
+		}
+		resolvedSubnets = append(resolvedSubnets, subnets...)
+	}
+
+	if len(resolvedSubnets) == 0 {
+		return nil, errors.New("unable to resolve at least one subnet")
+	}
+	if err := r.validateSubnetsAZExclusivity(resolvedSubnets); err != nil {
```

Comment:
We need to account for multiple matches per AZ, otherwise we make the tag selector feature unusable in those situations.

Reply:
How is this different from the existing annotation's tag searching semantics? The admin presumably has control over both the assignment of tags to subnets and use of tags in subnet selection; they can assign and select appropriately.

If we do anything other than error out, we risk painting ourselves into a corner should ALBs start supporting multiple subnets per AZ. We could perhaps add a check for this condition in the validating admission webhook so that it fails earlier.

Reply:
With `IngressClassParams`, one can also specify appropriate subnet IDs:

```
spec:
    subnets:
    - id: subnet-xxxx
    - tagKey: Name
      tagValue: mySubnet
 ```

to reproduce the example in the annotations documentation. (I'm thinking a followup PR could replace `SubnetsResolver.ResolveViaNameOrIDSlice()` with a call to `ResolveSubnetSpec()`.)

Reply:
If we provide semantics for multiple subnets in an AZ that are other than erroring out, then the semantics that ALB later gives multiple subnets in an AZ might turn out to not be backwards compatible with the semantics we provided.